hoomd/md/tune/nlist_buffer.py

```diff
@@ -0,0 +1,213 @@
+
+    def __call__(self):
+        if self._simulation.timestep == self._last_timestep:
+            return self._last_tps
+        walltime = self._simulation.walltime
+        timestep = self._simulation.timestep
+        if self._last_walltime is not None and self._last_timestep is not None:
+            delta_w = walltime - self._last_walltime
+            delta_t = timestep - self._last_timestep
+            # We divide by 1_000 to reduce the gradient size for optimization.
+            tps = delta_t / (1e3 * delta_w)
+        else:
+            tps = None
+        self._last_tps = tps
+        self._last_walltime = walltime
+        self._last_timestep = timestep
+        return tps
+
+
+class _NeighborListBufferInternal(hoomd.custom._InternalAction):
+    _skip_for_equality = {"_simulation", "_tunable"}
+
+    def __init__(
+        self,
+        nlist: NeighborList,
+        solver: hoomd.tune.solve.Optimizer,
+        maximum_buffer: float,
+    ):
+        self._simulation = None
+        self._tuned = False
+        self._max_tps = 0
+        self._best_buffer_size = None
+        param_dict = hoomd.data.parameterdicts.ParameterDict(
+            nlist=SetOnce(NeighborList),
+            solver=SetOnce(hoomd.tune.solve.Optimizer),
+            maximum_buffer=OnlyTypes(float,
+                                     postprocess=self._maximum_buffer_post))
+        param_dict.update({
+            "nlist": nlist,
+            "solver": solver,
+            "maximum_buffer": maximum_buffer
+        })
+        self._param_dict.update(param_dict)
+
+    def act(self, timestep: int):
+        if not self.tuned:
+            tps = self._tunable.y
+            if tps is not None and tps > self._max_tps:
+                self._best_buffer_size = self._tunable.x
+                self._max_tps = tps
+            self._tuned = self.solver.solve([self._tunable])
+
+    def _maximum_buffer_post(self, value: float):
+        if self._simulation is not None:
+            self._tunable.domain = (1e-5, value)
+        return value
+
+    def attach(self, simulation):
+        self._simulation = simulation
+        self._tunable = self._make_tunable(self.nlist)
+
+    def _make_tunable(self, nlist):
+        return hoomd.tune.ManualTuneDefinition(
+            get_y=_IntervalTPS(self._simulation),
+            target=0.0,
+            get_x=self._get_buffer,
+            set_x=self._set_buffer,
+            domain=(1e-5, self.maximum_buffer),
+        )
+
+    def _get_tps(self):
+        if self._simulation.tps == 0:
+            return None
+        return self._simulation.tps
+
+    def _get_buffer(self):
+        return self.nlist.buffer
+
+    def _set_buffer(self, new_buffer):
+        self.nlist.buffer = new_buffer
+
+    def detach(self):
+        self._simulation = None
+        self._tunable = None
+
+    @property
+    def tuned(self):
+        return self._tuned
+
+    @hoomd.logging.log
+    def max_tps(self):
+        return self._max_tps
```

Comment:
```suggestion
        return self._max_tps*1e3
```
Return TPS in units of timesteps per second.

Might be better to store `1e3` in a private module level constant.